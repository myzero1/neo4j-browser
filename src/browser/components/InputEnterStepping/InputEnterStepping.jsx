import { Component } from 'react'

export default class InputEnterStepping extends Component {
  state = {
    initial: true
  }
  steps = []
  getSubmitProps = () => {
    return {
      onClick: this.props.submitAction
    }
  }
  getInputPropsForIndex = (i, props = {}) => {
    this.steps[i] = this.steps[i] || {}
    const out = {
      ...props,
      onKeyDown: e => {
        // merge with users onKeyDown
        this.onKeyDown(e, i)
        props.onKeyDown && props.onKeyDown(e)
      }
    }
    if (props.initialFocus) {
      this.steps[i].focusOnRef = true
    }
    return out
  }
  onKeyDown = (e, i) => {
    if (e.keyCode === 13) {
      e.preventDefault()
      if (i !== this.steps.length - 1) {
        this.steps[i + 1].focusFn && this.steps[i + 1].focusFn()
      } else {
        this.props.submitAction && this.props.submitAction(e)
      }
    }
  }
  setRefForIndex = (i, ref) => {
    this.steps[i] = this.steps[i] || {}
    this.steps[i].focusFn = () => ref.focus()
    if (this.state.initial && this.steps[i].focusOnRef) {
      delete this.steps[i].focusOnRef
      this.setState({ initial: false }, () => {
        setTimeout(this.steps[i].focusFn, 0)
      })
    }
  }
  render () {
    return this.props.render({
      getInputPropsForIndex: this.getInputPropsForIndex,
      getSubmitProps: this.getSubmitProps,
      setRefForIndex: this.setRefForIndex
    })
  }
}
