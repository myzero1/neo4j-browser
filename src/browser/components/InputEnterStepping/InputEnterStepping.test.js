/* global jest */

import React from 'react'
import { render, Simulate } from 'react-testing-library'

import InputEnterStepping from './InputEnterStepping'

test('renders the render prop', () => {
  // When
  const { container } = render(
    <InputEnterStepping
      render={() => {
        return <div>hello</div>
      }}
    />
  )

  // Then
  expect(container).toMatchSnapshot()
})

test('submits on enter in last input', async () => {
  // Given
  const myFn = jest.fn()

  // When
  const { container, getByTestId } = render(
    <InputEnterStepping
      submitAction={myFn}
      render={({ getInputPropsForIndex, setRefForIndex }) => {
        return (
          <React.Fragment>
            <input
              {...getInputPropsForIndex(0, {
                'data-testid': 'first',
                defaultValue: 'first',
                ref: ref => setRefForIndex(0, ref)
              })}
            />
            <input
              {...getInputPropsForIndex(1, {
                'data-testid': 'second',
                defaultValue: 'second',
                ref: ref => setRefForIndex(1, ref)
              })}
            />
          </React.Fragment>
        )
      }}
    />
  )

  // Then
  expect(container).toMatchSnapshot()

  // When
  // Enter in first should focus second
  Simulate.keyDown(getByTestId('first'), {
    key: 'Enter',
    keyCode: 13,
    which: 13
  })

  // Then
  expect(myFn).toHaveBeenCalledTimes(0)

  // When
  // Enter in last should submit
  Simulate.keyDown(getByTestId('second'), {
    key: 'Enter',
    keyCode: 13,
    which: 13
  })

  // Then
  expect(myFn).toHaveBeenCalledTimes(1)
})

test('submits on button click', async () => {
  // Given
  const myFn = jest.fn()

  // When
  const { container, getByText } = render(
    <InputEnterStepping
      submitAction={myFn}
      render={({ getInputPropsForIndex, getSubmitProps, setRefForIndex }) => {
        return (
          <React.Fragment>
            <input
              {...getInputPropsForIndex(0, {
                'data-testid': 'first',
                defaultValue: 'first',
                ref: ref => setRefForIndex(0, ref)
              })}
            />
            <input
              {...getInputPropsForIndex(1, {
                'data-testid': 'second',
                defaultValue: 'second',
                ref: ref => setRefForIndex(1, ref)
              })}
            />
            <button {...getSubmitProps()}>Send</button>
          </React.Fragment>
        )
      }}
    />
  )

  // Then
  expect(container).toMatchSnapshot()
  expect(myFn).toHaveBeenCalledTimes(0)

  // When
  // Click button!
  Simulate.click(getByText('Send'))

  // Then
  expect(myFn).toHaveBeenCalledTimes(1)
})
