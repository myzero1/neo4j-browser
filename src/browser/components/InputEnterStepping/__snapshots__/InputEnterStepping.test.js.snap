// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the render prop 1`] = `
<div>
  <div>
    hello
  </div>
</div>
`;

exports[`submits on button click 1`] = `
<div>
  <input
    data-testid="first"
    value="first"
  />
  <input
    data-testid="second"
    value="second"
  />
  <button>
    Send
  </button>
</div>
`;

exports[`submits on enter in last input 1`] = `
<div>
  <input
    data-testid="first"
    value="first"
  />
  <input
    data-testid="second"
    value="second"
  />
</div>
`;
